### ISSUE TITLE
Watch mode failed to render

### ISSUE THREAD

**Issue by mdmoreau** [mohitsinghs/njk#10]
I'm pretty frequently getting messages like the following and an empty file when running in watch mode on Ubuntu 18.10:

```
ℹ Changed template src/html/index.html
⚠ Failed to render src/html/index.html, null output
✔ Wrote index.html
```

If I save the file again, it seems to build correctly. I also have the ```verbose``` flag enabled, but those are the only messages I'm seeing.

I think it might be related to https://github.com/babel/babel/issues/4332 and https://github.com/postcss/postcss-cli/issues/215, which were both fixed by adding ```awaitWriteFinish``` options to ```chokidar.watch```.

**Comment by mohitsinghs** [comment]
Just released v3.0.7, It fixes [two issues](https://github.com/mohitsinghs/njk/blob/master/CHANGELOG.md) one of which is related to this.

Outputs were null primarily because no actual error was handled while rendering a template, and more honestly errors were ignored with a null output. This is no longer the case and from now rendering error will fail with a stack trace.

This is better because we can take a look at the error to see if something is wrong in our config our njk is doing something unexpected.

Let's hope this will make this issue more easy to debug, before applying any fixes.

Please have a look at examples in the [wiki](https://github.com/mohitsinghs/njk/wiki) also.

Thanks for improving it 🎉.

**Comment by mdmoreau** [comment]
Thanks for the quick update! I'm unfortunately still having the same problem. The only difference is that now the ```⚠ Failed to render src/html/index.html, null output``` no longer appears, but an empty file is still generated. Saving again correctly builds the file most of the time.

I can sit on the same file without editing, and hit save every few seconds. Most of the time the file is generated correctly, but occasionally an empty file is created instead. The log always appears the same regardless of what happens, with no additional messages:

```
ℹ Changed template src/html/index.html
✔ Wrote index.html
```

Unfortunately adding awaitWriteFinish to chokidar doesn't seem to change anything for me locally.

**Comment by mohitsinghs** [comment]
Can you share me the file with the exact command you are running so that I can reproduce it.

**Comment by mdmoreau** [comment]
@mohitsinghs I setup a demo repo at https://github.com/mdmoreau/njk-demo. It doesn't happen super consistently, but it does output empty files a decent amount.

**Comment by mohitsinghs** [comment]
@mdmoreau Thanks for the setup. I tried it but it is quite surprising that I am unable reproduce the empty outputs despite of frequently saving both source files. Tested on windows too.

**Comment by mdmoreau** [comment]
@mohitsinghs Thanks for the help! This is what I'm getting in the log when an empty file gets created:

```
ℹ Changed template src/html/index.html
{ constructed: '' }
{ rendered: '' }
{ written: '' }
✔ Wrote index.html
```

**Comment by mohitsinghs** [comment]
> @mohitsinghs Thanks for the help! This is what I'm getting in the log when an empty file gets created:
> 
> ```
> ℹ Changed template src/html/index.html
> { constructed: '' }
> { rendered: '' }
> { written: '' }
> ✔ Wrote index.html
> ```

Looks like file is not even being passed. We need to dig deeper. Going update that branch with more logs right now.

**Comment by mdmoreau** [comment]
@mohitsinghs Thanks for the extra output. Getting this now when it builds an empty file:

```
{ chokidar: '/home/mdmoreau/Desktop/njk-demo/src/html/index.html' }
ℹ Changed template src/html/index.html
{ passed: '/home/mdmoreau/Desktop/njk-demo/src/html/index.html' }
{ preparing: '' }
{ constructed: '' }
{ rendered: '' }
{ written: '' }
✔ Wrote index.html
```

Here's the output when it renders correctly, though I imagine it's the same as what you're seeing:

```
{ chokidar: '/home/mdmoreau/Desktop/njk-demo/src/html/index.html' }
ℹ Changed template src/html/index.html
{ passed: '/home/mdmoreau/Desktop/njk-demo/src/html/index.html' }
{ preparing:
   '{% extends "templates/default.html" %}\n\n{% block body %}{% endblock %}\n' }
{ constructed:
   '{% extends "templates/default.html" %}\n\n{% block body %}{% endblock %}\n' }
{ rendered:
   '<!doctype html>\n<html lang="en">\n  <head>\n    <meta charset="utf-8">\n    <meta http-equiv="x-ua-compatible" content="ie=edge">\n    <title>build</title>\n    <meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1">\n  </head>\n  <body>\n    \n  </body>\n</html>\n' }
{ written:
   '<!doctype html>\n<html lang="en">\n  <head>\n    <meta charset="utf-8">\n    <meta http-equiv="x-ua-compatible" content="ie=edge">\n    <title>build</title>\n    <meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1">\n  </head>\n  <body>\n    \n  </body>\n</html>\n' }
✔ Wrote index.html
```

This is when saving through VS Code on Ubuntu 18.10 for what it's worth.

**Comment by mdmoreau** [comment]
@mohitsinghs I figured I'd try saving through another editor (gedit), and I haven't been able to generate an empty file yet. The strange thing is that I'm getting all of this output on just one file save:

```
{ chokidar:
   '/home/mdmoreau/Desktop/njk-demo/src/html/.goutputstream-TJ1LUZ' }
ℹ Changed template src/html/.goutputstream-TJ1LUZ
{ passed: '/home/mdmoreau/Desktop/njk-demo/src/html/index.html' }
{ chokidar: '/home/mdmoreau/Desktop/njk-demo/src/html/index.html' }
ℹ Changed template src/html/index.html
{ passed: '/home/mdmoreau/Desktop/njk-demo/src/html/index.html' }
{ preparing:
   '{% extends "templates/default.html" %}\n\n{% block body %}{% endblock %}\n' }
{ constructed:
   '{% extends "templates/default.html" %}\n\n{% block body %}{% endblock %}\n' }
{ rendered:
   '<!doctype html>\n<html lang="en">\n  <head>\n    <meta charset="utf-8">\n    <meta http-equiv="x-ua-compatible" content="ie=edge">\n    <title>build</title>\n    <meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1">\n  </head>\n  <body>\n    \n  </body>\n</html>\n' }
{ written:
   '<!doctype html>\n<html lang="en">\n  <head>\n    <meta charset="utf-8">\n    <meta http-equiv="x-ua-compatible" content="ie=edge">\n    <title>build</title>\n    <meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1">\n  </head>\n  <body>\n    \n  </body>\n</html>\n' }
{ preparing:
   '{% extends "templates/default.html" %}\n\n{% block body %}{% endblock %}\n' }
{ constructed:
   '{% extends "templates/default.html" %}\n\n{% block body %}{% endblock %}\n' }
{ rendered:
   '<!doctype html>\n<html lang="en">\n  <head>\n    <meta charset="utf-8">\n    <meta http-equiv="x-ua-compatible" content="ie=edge">\n    <title>build</title>\n    <meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1">\n  </head>\n  <body>\n    \n  </body>\n</html>\n' }
{ written:
   '<!doctype html>\n<html lang="en">\n  <head>\n    <meta charset="utf-8">\n    <meta http-equiv="x-ua-compatible" content="ie=edge">\n    <title>build</title>\n    <meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1">\n  </head>\n  <body>\n    \n  </body>\n</html>\n' }
✔ Wrote index.html
✔ Wrote index.html
{ chokidar:
   '/home/mdmoreau/Desktop/njk-demo/src/html/.goutputstream-TJ1LUZ' }
ℹ Changed template src/html/.goutputstream-TJ1LUZ
{ passed: '/home/mdmoreau/Desktop/njk-demo/src/html/index.html' }
{ preparing:
   '{% extends "templates/default.html" %}\n\n{% block body %}{% endblock %}\n' }
{ constructed:
   '{% extends "templates/default.html" %}\n\n{% block body %}{% endblock %}\n' }
{ rendered:
   '<!doctype html>\n<html lang="en">\n  <head>\n    <meta charset="utf-8">\n    <meta http-equiv="x-ua-compatible" content="ie=edge">\n    <title>build</title>\n    <meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1">\n  </head>\n  <body>\n    \n  </body>\n</html>\n' }
{ written:
   '<!doctype html>\n<html lang="en">\n  <head>\n    <meta charset="utf-8">\n    <meta http-equiv="x-ua-compatible" content="ie=edge">\n    <title>build</title>\n    <meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1">\n  </head>\n  <body>\n    \n  </body>\n</html>\n' }
✔ Wrote index.html
```

It seems like that is working correctly, though it looks odd, and the issue might be with VS Code (https://github.com/Microsoft/vscode/issues/61310).

**Comment by mohitsinghs** [comment]
@mdmoreau Extra output is due to extra events emitted by chokidar which is due to swap files created by editors likes vim or gedit. I'll ignore dotfiles in the next release to fix that.

**Comment by mohitsinghs** [comment]
I made some more changes in `null output` branch and we are watching for changes only ( instead of all events ).
This  should fix extra output for editors like vim and gedit.

**Comment by mdmoreau** [comment]
@mohitsinghs Seeing normal output when using gedit now, but still getting the following with the frequent empty file generation in VS Code:

```
{ chokidar: '/home/mdmoreau/Desktop/njk-demo/src/html/index.html' }
ℹ Changed template src/html/index.html
{ passed: '/home/mdmoreau/Desktop/njk-demo/src/html/index.html' }
{ preparing: '/home/mdmoreau/Desktop/njk-demo/src/html/index.html' }
{ nonpromised: '' }
{ promised: '' }
{ constructed: '' }
{ rendered: '' }
{ written: '' }
✔ Wrote index.html
```

May be worth just letting it sit for a while since it seems like it might be an issue with VS Code. Appreciate the quick responses though!

**Comment by mohitsinghs** [comment]
From this test this is confirm now that file is empty and some other tool is doing that ( Might be VSCode or some plugin of it ).

**Comment by mdmoreau** [comment]
Sounds good - thanks!

**Comment by mohitsinghs** [comment]
@mdmoreau Just released v3.0.8 🎉
Problems related to extra outputs for swap files are fixed now. Also, marked is updated to latest version.
